pkg/controller/deployment/deployment_controller.go

```diff
@@ -24,51 +24,291 @@ import (
+	// TODO: Is this the right way to check this, or is checking names sufficient?
+	if !api.Semantic.DeepEqual(oldRC, curRC) {
+		if oldD := dc.getDeploymentForRC(oldRC); oldD != nil {
+			dc.enqueueDeployment(oldD)
 		}
 	}
-	return nil
 }
 
-func (d *DeploymentController) reconcileDeployment(deployment *experimental.Deployment) error {
-	targetedRCs, err := d.getTargetedRCs(deployment)
+// When a controller is deleted, enqueue the deployment that manages it.
+// obj could be an *api.ReplicationController, or a DeletionFinalStateUnknown
+// marker item.
+func (dc *DeploymentController) deleteRC(obj interface{}) {
+	rc, ok := obj.(*api.ReplicationController)
+
+	// When a delete is dropped, the relist will notice a pod in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the RC
+	// changed labels the new deployment will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+		rc, ok = tombstone.Obj.(*api.ReplicationController)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not an rc %+v, could take up to %v before a deployment recreates/updates controllers", obj, FullDeploymentResyncPeriod)
+			return
+		}
+	}
+	if d := dc.getDeploymentForRC(rc); d != nil {
+		dc.enqueueDeployment(d)
+	}
+}
+
+// obj could be an *api.Deployment, or a DeletionFinalStateUnknown marker item.
+func (dc *DeploymentController) enqueueDeployment(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
 	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	// TODO: Handle overlapping deployments better. Either disallow them at admission time or
+	// deterministically avoid syncing deployments that fight over RC's. Currently, we only
+	// ensure that the same deployment is synced for a given RC. When we periodically relist
+	// all deployments there will still be some RC instability. One way to handle this is
+	// by querying the store for all deployments that this deployment overlaps, as well as all
+	// deployments that overlap this deployments, and sorting them.
+	dc.queue.Add(key)
+}
+
+func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go dc.dController.Run(stopCh)
+	go dc.rcController.Run(stopCh)
+	go dc.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(dc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down deployment controller")
+	dc.queue.ShutDown()
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
```

Comment:
This assertion is true in-process, but it seems important to point out that `syncHandler` must still behave correctly during concurrent processing of the same key (consider multiple active deployment controller processes running within the same cluster.)